### PR TITLE
Theme Upload: Improve the notifications for AT transfer with theme upload

### PR DESCRIPTION
**Pull request by BogdanUngureanu**
This PR improves the AT notifications and transfer errors for the theme upload interface. 

Previously, the user was not presented with any input (or was behaving incorrectly) that signals:
- An error has occurred while setting up the theme
- The theme upload & AT transfer was finished

Additionally, I also removed the WP-Admin redirect on Atomic because it creates an on-optimal user experience. If the transfer is completed (or failed) and the user refreshes the page, they would be redirected to a totally different UI. 

Related to https://github.com/Automattic/wp-calypso/issues/93581 and https://github.com/Automattic/wp-calypso/issues/93582

## Proposed Changes

* Display correctly the in-progress notice
* Display an error on theme configuration
* Remove the WP-Admin redirect on AT

## Why are these changes being made?

* bug fix

## Testing Instructions

* Apply D160747-code on your sandbox
* Use the live link and go to /themes/upload/your-business-plan-site
* Upload a zip file with a theme
* You should get the in-progress notice
* The theme UI card should be displayed after the upload 

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/state/themes/actions/theme-transfer.js [hunk 1]
```diff
@@ -170,6 +171,11 @@ export function pollThemeTransferStatus( siteId, transferId, interval = 3000, ti
 				.then( ( { status, message, uploaded_theme_slug } ) => {
 					dispatch( transferStatus( siteId, transferId, status, message, uploaded_theme_slug ) );
 					if ( status === 'complete' ) {
+						dispatch( {
+							type: THEME_UPLOAD_SUCCESS,
+							siteId,
+							themeId: uploaded_theme_slug,
+						} );
```
Comment (mmtr) [inline client/state/themes/actions/theme-transfer.js]: Is this a right assumption? I _think_ `pollThemeTransferStatus` can run in other contexts than uploading a theme because it's part of the `initiateThemeTransfer` action:

https://github.com/Automattic/wp-calypso/blob/66c1aea529d734ee3cab9f5a1a877bd2ee5b963e/client/state/themes/actions/theme-transfer.js#L101

- Activate hosting features:

https://github.com/Automattic/wp-calypso/blob/66c1aea529d734ee3cab9f5a1a877bd2ee5b963e/client/my-sites/hosting/hosting-activate.js#L49

- Install marketplace product:

https://github.com/Automattic/wp-calypso/blob/66c1aea529d734ee3cab9f5a1a877bd2ee5b963e/client/my-sites/marketplace/pages/marketplace-product-install/index.tsx#L217

File: client/state/themes/test/actions.js [hunk 2]
```diff
@@ -1120,7 +1120,7 @@ describe( 'actions', () => {
 		test( 'should dispatch status update', async () => {
 			await pollThemeTransferStatus( siteId, 3, 20 )( spy );
 			// Two 'progress' then a 'complete'
-			expect( spy ).toHaveBeenCalledTimes( 3 );
+			expect( spy ).toHaveBeenCalledTimes( 4 );
 			expect( spy ).toHaveBeenCalledWith( {
```
Comment (mmtr) [inline client/state/themes/test/actions.js]: We should update the comment above too. I guess now it's "Two 'progress' then a 'complete' _and something else_"

File: client/state/themes/upload-theme/reducer.js [hunk 3]
```diff
@@ -247,11 +247,45 @@ export const isTransferComplete = ( state = {}, action ) => {
...
+
+			return {
+				...state,
+				[ siteId ]: status !== 'complete',
+			};
+		}
+	}
+
+	return state;
+};
```
Comment (mmtr) [inline client/state/themes/upload-theme/reducer.js:289]: Why this new reducer and the associated new selector? Cannot we re-use something from the existing state like `isAutomatedTransferActive`?

https://github.com/Automattic/wp-calypso/blob/66c1aea529d734ee3cab9f5a1a877bd2ee5b963e/client/state/automated-transfer/selectors/is-automated-transfer-active.js#L30

File: client/data/themes/with-is-fse-active.js [hunk 4]
```diff
@@ -11,7 +11,14 @@ const withIsFSEActive = createHigherOrderComponent(
 		const { data, isLoading } = useActiveThemeQuery( siteId, userLoggedIn );
 		const isFSEActive = data?.[ 0 ]?.is_block_theme ?? false;
 
-		return <Wrapped { ...props } isFSEActiveLoading={ isLoading } isFSEActive={ isFSEActive } />;
+		return (
+			<Wrapped
+				{ ...props }
+				isFSEActiveLoading={ isLoading }
+				activeTheme={ data?.[ 0 ]?.stylesheet }
```
Comment (mmtr) [inline client/data/themes/with-is-fse-active.js]: Can this potentially conflict with other props like https://github.com/Automattic/wp-calypso/blob/66c1aea529d734ee3cab9f5a1a877bd2ee5b963e/client/my-sites/themes/single-site.jsx#L50?

I think it might be safer to add a new `activeTheme` prop to the specific components where is needed (I think it's only needed in `Upload`,  right?).

File: client/my-sites/themes/theme-upload/index.jsx [hunk 5]
```diff
@@ -435,13 +428,17 @@ class Upload extends Component {
 				></NavigationHeader>
 
 				<HeaderCake backHref={ backPath }>{ translate( 'Install theme' ) }</HeaderCake>
-				{ ! showTrialAcknowledgeModal && ! isAtomic && (
+				{ ! showTrialAcknowledgeModal && shouldShowAtNotice && (
 					<HostingActivateStatus
 						context="theme"
 						onTick={ this.requestUpdatedSiteData }
 						keepAlive={ hasRequestedTrial && ! isAtomic }
+						forceEnable={ shouldShowAtNotice }
```
Comment (mmtr) [inline client/my-sites/themes/theme-upload/index.jsx]: Minor (non-blocking) but could we check directly `this.props.isTransferInProgress` here instead of using a var with a different name to have consistency with the `this.props.isTransferCompleted` check below?

**Review by mmtr — COMMENTED** [review]
This is looking good! Left some suggestions and questions

File: client/state/themes/upload-theme/reducer.js (hunk 3 above)
Comment (BogdanUngureanu) [reply client/state/themes/upload-theme/reducer.js:289]: We need the new reducer because inProgress also contains the events for the theme upload, the logic that handles uploads on sites that are Atomic. 

In this case, it would mean the "Please wait until we enable the theme features" notice would display for regular uploads as well (although we do already have a progress bar ui).

Comment (mmtr) [reply client/state/themes/upload-theme/reducer.js:289]: > inProgress also contains the events for the theme upload

Sorry, I don't follow. I wasn't asking why we cannot re-use the `inProgress` prop.

I was asking why we add a new reducer and a new selector that checks whether the transfer is in progress, when we already have existing utils like `isAutomatedTransferActive`.

Comment (mmtr) [reply client/state/themes/upload-theme/reducer.js:289]: Related: I see the component has a `isTransferring` state. Cannot we use that as well?

Comment (BogdanUngureanu) [reply client/state/themes/upload-theme/reducer.js:289]: >Sorry, I don't follow. I wasn't asking why we cannot re-use the inProgress prop.

Oops, I should probably stop responding to PR comments at 9 pm 😅 

> I was asking why we add a new reducer and a new selector that checks whether the transfer is in progress, when we already have existing utils like isAutomatedTransferActive.

The ThemeUpload component uses `initiateThemeTransfer` that dispatches its own transfer status events (e.g. `THEME_TRANSFER_INITIATE_REQUEST`) and stores them in a separate state. I'm not sure if `isAutomatedTransferActive` would work, but even if it works, it would mean having the status extracted from multiple places. 

> Related: I see the component has a isTransferring state. Cannot we use that as well?

That prop seems to be populated from `HostingActivateStatus` component; which for some reason doesn't seem to work nice with the Theme Upload implementation.

Comment (mmtr) [reply client/state/themes/upload-theme/reducer.js:289]: > `initiateThemeTransfer` that dispatches its own transfer status events

Seems that the `automated-transfer` state use the same actions:

https://github.com/Automattic/wp-calypso/blob/4b4c0f9e7c12d7e639a8ef06ffd1e4648c75ef04/client/state/automated-transfer/reducer.js#L9-L11

I have the feeling we're duplicating the code here.

Comment (BogdanUngureanu) [reply client/state/themes/upload-theme/reducer.js:289]: Yeah, it's duplicated. I'll give it a try to re-implement it to use the generic AT logic. I didn't want to change the current architecture, but it would probably make the code more cleaner.

Comment (BogdanUngureanu) [reply client/state/themes/upload-theme/reducer.js:289]: I've tried using that selector instead but the implementation is not working as expected: the notice ui appears for a second and dissapears and re-appears after 20s. This is because the status is [refetched](https://github.com/Automattic/wp-calypso/blob/4b55037b44daee8daacd0f4b49937f687e0b62a7/client/my-sites/hosting/hosting-activate-status.tsx#L69) immediately, before the AT transfer actually starts:
- initiate request is made (request pending) -> AT Status: start
- the notice appears
- An AT fetch status is requested - the AT process hasn't started -> new status none
- The notice dissapears
- initiate request is finished (usually takes 10-15-20 seconds) - it's a file upload after all
- A fetch status is requested -> new status is now provisioning
- The notice appears again

We don't have this problem with the theme AT state because it behaves differently, it's not using only the AT revert states to determine if the AT is in progress. 

Implementing a proper solution in the automated-transfer state would require a bit more effort and I'm not 100% sure we won't introduce regressions.

**Review by mmtr — APPROVED** [review]
Looks good!

> The upload should fail (since the back-end issue will be fixed by the Atomic team) see pMz3w-kCQ-p2

Can we maybe wait before deploying these changes until the Atomic team fixes the issue? Just to be sure that these changes work with their fix.

**Comment by BogdanUngureanu** [comment]
> Can we maybe wait before deploying these changes until the Atomic team fixes the issue? Just to be sure that these changes work with their fix.

We can wait, but this would still be an improvement to what we have right now. if the user uses the UI, they'll be notified that the theme upload failed. If they try again to upload the file, it'll work because the site is already AT and a separate API is used to upload the theme.

Fyi. I'm still working on fixing some small UI glitches when the AT with theme upload is working (with the hardcoded API change). 😅

**Comment by mmtr** [comment]
@BogdanUngureanu Feel free to ship it then!

**Comment by BogdanUngureanu** [comment]
@mmtr can you give it another spin?

File: client/state/themes/actions/theme-transfer.js [hunk 6]
```diff
@@ -153,11 +153,18 @@ function transferInitiateFailure( siteId, error, plugin, context ) {
...
+ * @param {boolean} hasThemeFileUpload -- has theme File Upload
  * @param {number} [interval] -- time between poll attempts
  * @param {number} [timeout] -- time to wait for 'complete' status before bailing
  * @returns {Promise} for testing purposes only
  */
-export function pollThemeTransferStatus( siteId, transferId, interval = 3000, timeout = 180000 ) {
+export function pollThemeTransferStatus(
+	siteId,
+	transferId,
+	hasThemeFileUpload = false,
```
Comment (mmtr) [inline client/state/themes/actions/theme-transfer.js]: Can this new param be the last param? Unless we're 100% sure that every single place calling this function is using the new signature.

Comment (BogdanUngureanu) [reply client/state/themes/actions/theme-transfer.js]: Done :)

**Comment by mmtr** [comment]
The error notice now disappears after a moment, but it was permanent before:


https://github.com/user-attachments/assets/68d0cfdf-b0b7-40dd-967e-a773b4313526

**Comment by BogdanUngureanu** [comment]
I've pushed a new version that simplifies things a bit, and makes it more consistent with the error notices we already use on AT theme upload scenario.

So, instead of displaying the error notice in the form, we now use the drawer notice (top-right of the screen).

Additionally, I've updated the testing steps to include a WPCOM diff that needs to be applied on the sandbox for the Calypso implementation to work.

File: client/data/themes/use-active-theme-query.ts [hunk 7]
```diff
@@ -23,12 +23,19 @@ export const useActiveThemeQuery = (
...
 				apiNamespace: 'wp/v2',
 			} );
+
+			if ( ! Array.isArray( data ) || data.length === 0 ) {
+				throw new Error( 'Data is empty or not an array' );
+			}
+
+			return data;
 		},
+		retry: 10,
```
Comment (mmtr) [inline client/data/themes/use-active-theme-query.ts]: Are these changes needed now after your latest refactor?

File: client/components/marketing-survey/cancel-purchase-form/style.scss [hunk 8]
```diff
@@ -1,5 +1,6 @@
 @import "@wordpress/base-styles/breakpoints";
 @import "@wordpress/base-styles/mixins";
+@import "@wordpress/base-styles/colors";
```
Comment (mmtr) [inline client/components/marketing-survey/cancel-purchase-form/style.scss]: Why is this needed?

File: client/my-sites/themes/theme-upload/index.jsx [hunk 9]
```diff
@@ -312,9 +300,11 @@ class Upload extends Component {
 				<div className="theme-upload__description">{ theme.description }</div>
 				<div className="theme-upload__action-buttons">
 					<Button onClick={ this.onTryAndCustomizeClick }>{ tryandcustomize.label }</Button>
-					<Button primary onClick={ this.onActivateClick }>
-						{ activate.label }
-					</Button>
+					{ this.props.activeTheme !== theme.id && ! this.props.isTransferCompleted && (
```
Comment (mmtr) [inline client/my-sites/themes/theme-upload/index.jsx]: This looks wrong? We shouldn't allow the theme to be activated if the transfer has not been completed, but here, we're allowing it only if the transfer is still in progress 🤔 Am I missing anything?

File: client/my-sites/themes/theme-upload/index.jsx [hunk 10]
```diff
@@ -435,11 +425,12 @@ class Upload extends Component {
 				></NavigationHeader>
 
 				<HeaderCake backHref={ backPath }>{ translate( 'Install theme' ) }</HeaderCake>
-				{ ! showTrialAcknowledgeModal && ! isAtomic && (
+				{ ! showTrialAcknowledgeModal && this.props.isTransferInProgress && (
 					<HostingActivateStatus
 						context="theme"
 						onTick={ this.requestUpdatedSiteData }
 						keepAlive={ hasRequestedTrial && ! isAtomic }
+						forceEnable={ this.props.isTransferInProgress }
```
Comment (mmtr) [inline client/my-sites/themes/theme-upload/index.jsx]: `keepAlive` and `forceEnable` seem to have the same purpose. Can you join them?
```suggestion
						keepAlive={ ( hasRequestedTrial && ! isAtomic ) || this.props.isTransferInProgress }
```

File: client/state/themes/actions/theme-transfer.js [hunk 11]
```diff
@@ -154,11 +155,40 @@ function transferInitiateFailure( siteId, error, plugin, context ) {
...
+						siteId,
+						themeId: uploaded_theme_slug,
+					} );
+					// finished, stop polling
+					return resolve();
+				}
+
+				if ( Date.now() > endTime ) {
+					return setTimeout( tryThemeFetch, 1000, uploaded_theme_slug, resolve );
+				}
```
Comment (mmtr) [inline client/state/themes/actions/theme-transfer.js]: Am I reading this wrong, or can this cause `tryThemeFetch` to run infinitely if the theme never gets activated after exceeding the allowed timeout?

**Review by mmtr — COMMENTED** [review]
> I've updated the testing steps to include a WPCOM diff that needs to be applied on the sandbox for the Calypso implementation to work.

Asked in the diff too, but should we expect then the Atomic transfer to work after uploading a theme? I still get a failure:


https://github.com/user-attachments/assets/ac2976ad-5cca-4981-a649-38163302b1eb

**Comment by mmtr** [comment]
> 
Asked in the diff too, but should we expect then the Atomic transfer to work after uploading a theme? I still get a failure

Works now after the last update to that diff 🎉

File: client/state/themes/actions/theme-transfer.js (hunk 11 above)
Comment (BogdanUngureanu) [reply client/state/themes/actions/theme-transfer.js]: endTime condition should prevent that.

File: client/components/marketing-survey/cancel-purchase-form/style.scss (hunk 8 above)
Comment (BogdanUngureanu) [reply client/components/marketing-survey/cancel-purchase-form/style.scss]: Oupsy. Removed it, thanks!

File: client/my-sites/themes/theme-upload/index.jsx (hunk 10 above)
Comment (BogdanUngureanu) [reply client/my-sites/themes/theme-upload/index.jsx]: yeah, they are similar, but `keepAlive` is guarded by the `isTransferCompleted` constant from `HostingActivateStatus`. With this new prop we are forcing the state instead of using the one from `HostingActivateStatus`. 

I've contemplated using directly the `Notice` component, but I thought it makes sense to use the same  `HostingActivateStatus` component everywhere (it should be easier to refactor changes in the future this way) so that we can get a list of all places where AT notices are displayed.

File: client/my-sites/themes/theme-upload/index.jsx (hunk 9 above)
Comment (BogdanUngureanu) [reply client/my-sites/themes/theme-upload/index.jsx]: isTransferCompleted, in context of theme upload, means that the theme was also installed on the site. The complete action is triggered only after the polling logic determines that the theme is active on the site.

**Review by mmtr — APPROVED** [review]
